include/plat/spike/plat/machine/hardware.h

```diff
@@ -36,20 +46,33 @@
  * as the largest frame size */
 #define PADDR_LOAD 0xC0000000lu
 
+#ifdef CONFIG_SEL4_RV_MACHINE
+#define PADDR_TOP -1UL
+#else
 /* The highest valid physical address that can be indexed in the kernel window */
 #define PADDR_TOP (KERNEL_BASE - PPTR_BASE + PADDR_BASE)
+#endif
+
 /* The highest valid physical address that can be used for the kernel image. We offset by
  * PADDR_LOAD as the window for the kernel image is mapped started at PADDR_LOAD */
 #define PADDR_HIGH_TOP (-KERNEL_BASE + PADDR_LOAD)
 
-/* Translates from a physical address and a value in the kernel image */
-#define KERNEL_BASE_OFFSET (KERNEL_BASE - PADDR_LOAD)
-
 /* Convert our values into general values expected by the common code */
 #define kernelBase KERNEL_BASE
+
+#ifdef CONFIG_SEL4_RV_MACHINE
+#define BASE_OFFSET 0
+#define KERNEL_BASE_OFFSET 0
+#define PPTR_TOP PADDR_TOP
+#define PPTR_USER_TOP PADDR_TOP
+#else
 #define PPTR_TOP KERNEL_BASE
 #define PPTR_USER_TOP PPTR_BASE
 #define BASE_OFFSET (PPTR_BASE - PADDR_BASE)
+/* Translates from a physical address and a value in the kernel image */
+#define KERNEL_BASE_OFFSET (KERNEL_BASE - PADDR_LOAD)
```

Comment:
I don't understand why this `#define` is special cased. Doing so implies that `KERNEL_BASE - PADDR_LOAD` is *not* going to be 0 in M-mode, but surely this can only be correct if `KERNEL_BASE` and `PADDR_LOAD` are the same value? Am I missing something in the memory layout for the kernel in M-mode?

Reply:
You are right, it's 0 in machine mode. I just added it to avoid confusion (and also for someone who's reading the code). Furthermore, the comment doesn't apply in machine mode as well, so I decided to make them explicitly separate.

Reply:
I think the comment does apply. Code in the kernel generally isn't M-mode specific and so does need to translate from a physical address to the kernel image virtual address, which is what the `KERNEL_BASE_OFFSET` macro does. The translation being the identify function in the case of M-mode does not make it not a translation.

Having separate definitions is more confusing as it forces the question of why do they need to defined separately.

Similarly I would define a correct value for `PPTR_BASE`, not just because it allows for a consistent `BASE_OFFSET` definition to be used, but also this is a value that has meaning and should be defined correctly. Just because it is currently only used in kernel window mapping code, which isn't being executed, doesn't mean it shouldn't be defined correctly.